### ISSUE TITLE
Create documentation on how to test extensions

### ISSUE THREAD

**Issue by txomon** [mopidy/mopidy#503]
This would be a quite welcomed documentation improvement. I have finally worked it out using tests/backends/local/* files to find out the way, but may not be straight-forward without help.

Thus, adding some text about testing to the docs on the Extension development chapter would be great!

**Comment by jcass77** [comment]
I also found testing to be quite difficult to get into. To a large extent the approach depends on how your extensions are structured, which parts of Mopidy it interacts with, and if it uses any 3rd party APIs or makes any HTTP requests to the outside world.

Based on my own experience, I think some general pointers might be really useful for newcomers.

I can perhaps work on creating documentation outlining the broad approach for testing if there is general agreement that what is being suggested below is sensible (semi-pseudo code, might not be valid/working):

**Extensions:**

As a minimum general tests for extensions should ensure that:
- it provides a default configuration in ```ext.conf``` with the correct parameters and values,
- the config schema is complete,
- it is added to the Mopidy registry on setup,

**Backends:**

Backends can usually be constructed with a small mockup of the configuration file, and mocking the audio actor:

```python
@pytest.fixture()
def config():
    return {
        'http': {
            'hostname': '127.0.0.1',
            'port': '6680'
        },
        'proxy': {
            'hostname': 'host_mock',
            'port': 'port_mock'
        },
        'my_extension': {
            'enabled': True,
            'param_1': 'value_1',
            'param_2': 'value_2',
            'param_n': 'value_n',
        }
    }

def get_backend(config):
    return backend.MyBackend(config=config, audio=mock.Mock())
```

You'll probably want to patch ```requests``` or any other web API's that you use to avoid any unintended HTTP requests from being made by your backend during testing:

```python
from mock import patch 
@mock.patch('requests.get', mock.Mock(side_effect=<either raise a specific exception that you want to test responses on or a general exception to avoid unintended HTTP requests>)
```

As a minimum your backend tests should cover that:
- it provides a unique URI scheme,
- it sets up the various providers (e.g. library / playback)

Once you have a backend instance to work with, testing the various playback, library, and other providers is straight forward and should not require any special setup / processing.

**Libraries:**

Tests for these should just cover the standard Mopidy API (```browse```, ```lookup```, ```refresh```, ```get_images```, ```search```, etc.)

**Playback:**

Testing ```change_track``` and ```translate_uri``` is probably the highest priority.

**Frontends:**

For me testing frontends are the most difficult / confusing. Because most frontends will interact with the Mopidy core, I found that I needed to have one running to be able to test anything:

```python
self.core = core.Core.start(
            config, backends=[get_backend(config]).proxy()
```

You'll also need a list of ```models.Track``` and a list of uris in order to populate the core with tracks:

```python
class BaseTest(unittest.TestCase):
    tracks = [
        models.Track(uri='my_scheme:track:id1', length=40000),  # Regular track
        models.Track(uri='my_scheme:track:id2', length=None),   # No duration
    ]

    uris = [ 'my_scheme:track:id1', 'my_scheme:track:id2']
```
In the ```setup()``` method of your test class, you will then probably need to monkey patch looking up tracks in the library (so that it will always use the lists that you defined), and then populate core's tracklist:

```python
def lookup(uris):
    result = {uri: [] for uri in uris}
    for track in self.tracks:
        if track.uri in result:
            result[track.uri].append(track)
    return result

self.core.library.lookup = lookup
self.tl_tracks = self.core.tracklist.add(uris=self.uris).get()
```

With all of that done you should finally be ready to instantiate your frontend:

```self.frontend = frontend.MyFrontend.start(config(), self.core).proxy()```

...and then just remember that the normal core and frontend methods will usually return ```pykka.ThreadingFuture``` objects, so you will need to add ```.get()``` at the end of pretty much everything to get the actual return values.

Some of the more recent code in the develop branch also contains neat examples for triggering events that your frontend can respond on:

```python
self.events = []
self.patcher = mock.patch('mopidy.listener.send')
self.send_mock = self.patcher.start()

def send(cls, event, **kwargs):
    self.events.append((event, kwargs))

self.send_mock.side_effect = send
```
...and then just call ```replay_events()``` at the relevant points in your code to have the events fire:

```python
def replay_events(self, frontend, until=None):
    while self.events:
        if self.events[0][0] == until:
            break
        event, kwargs = self.events.pop(0)
        frontend.on_event(event, **kwargs).get()
```

**Comment by jcass77** [comment]
I should add that the one thing I haven't been able to figure out is how to check if a method was called in the core or the frontend.

Just doing ```fronted.method = mock.Mock()``` and then checking with ```assert fronted.method.called``` doesn't work. I think it is due to the way that the pykka threading works, and am not sure how to do these checks in a pykka environment - any help would be much appreciated.

**Comment by jodal** [comment]
Without answering your question about testing if core methods are called, I'd like to say that what you have here looks very good, and should clearly go into the docs.

I haven't been very active lately, but I've noticed a lot of good contributions from you, so I'd like to make you a contributor both to give credit for your contributions and to make you able to assign issues to yourself. Thanks! :-)

**Comment by jodal** [comment]
Fixed by PR #1401.
